### PR TITLE
 FrozenSmt not holding oldest ancester 

### PR DESCRIPTION
**Pull request by msmouse**
### Description
Now that ledger update is a seperate stage, the oldest ancester is held                                                                                                                                  from multiple threads, which can result in it not being freed when the                                                                                                                                   pipeline is busy.    

### Test Plan
On certain machines where we used to observe high memory usage, it's now good.
Local benchmark runs didn't observe regression.
Unit tests.

### CODE REVIEW

File: storage/scratchpad/src/sparse_merkle/ancestors.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+type Result<V> = std::result::Result<V, Error>;
+
+/// Keep the oldest SMTs in a central place so that:
+/// 1. elsewhere, dropping the SMT will be fast since this indirectly hols a ref to every tree in
+///    the entire forest.
+/// 2. this must be invoked somewhere in the critical path so that it can provide some
+///    back pressure to slow things down to prevent from leaking memory
+#[derive(Clone, Debug)]
+pub struct SmtAncestors<V: Clone + Send + Sync + 'static> {
```
Comment (grao1991) [inline storage/scratchpad/src/sparse_merkle/ancestors.rs:26]: nit: remove Clone

**Review by grao1991 — APPROVED** [review]

File: storage/scratchpad/src/sparse_merkle/ancestors.rs (hunk 1 above)
Comment (msmouse) [reply storage/scratchpad/src/sparse_merkle/ancestors.rs:26]: Clone is required down the stack from the updater. I'd follow up later on this in favor of time. 😢

File: storage/scratchpad/src/sparse_merkle/ancestors.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            oldest.log_generation("evict_ancestor");
+            if !oldest.is_the_only_ref() {
+                sample!(
+                    SampleRate::Duration(Duration::from_secs(1)),
+                    error!(
+                        "Oldest SMT being tracked by SmtAncestors is still referenced elsewhere. Potential memory leak.",
+                    )
+                );
+            } else {
+                self.drop_queue.enqueue_drop(oldest);
```
Comment (areshand) [inline storage/scratchpad/src/sparse_merkle/ancestors.rs:77]: is it supposed to enque when MAX_PENDING_DROPS reaches?

Comment (msmouse) [reply storage/scratchpad/src/sparse_merkle/ancestors.rs:77]: enqueue_drop

Comment (msmouse) [reply storage/scratchpad/src/sparse_merkle/ancestors.rs:77]: `enqueue_drop()` will block until backlog drops back below that

File: storage/aptosdb/src/test_helper.rs [hunk 3]
```diff
@@ -109,20 +108,16 @@ pub fn update_in_memory_state(state: &mut StateDelta, txns_to_commit: &[Transact
         if txn_to_commit.is_state_checkpoint() {
             state.current = state
                 .current
-                .clone()
-                .freeze()
```
Comment (areshand) [inline storage/aptosdb/src/test_helper.rs:113]: why is this freeze and unfreeze removed?

**Review by areshand — APPROVED** [review]

File: storage/aptosdb/src/test_helper.rs (hunk 3 above)
Comment (msmouse) [reply storage/aptosdb/src/test_helper.rs:113]: just a different way to hack.. (SparseMerkleTree::batch_update() now generates `Usage::Untracked` internally, as is the intention here)

**Pull request merged**